### PR TITLE
[libxslt] add optional feature support 

### PR DESCRIPTION
**Pull request by GordonSmith**
Expose some libxslt optional features

- #### What does your PR fix?
  NA

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  No change to triplet supprt

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

File: ports/libxslt/vcpkg.json [hunk 1]
```diff
@@ -1,13 +1,30 @@
...
+      "description": "Build with crypto support (--with-crypto), !windows",
+      "dependencies": [
+        "libgcrypt"
+      ]
+    },
+    "plugins": {
+      "description": "Build with plugins (--with-plugins), !windows"
+    },
+    "python": {
+      "description": "Builds with python support (--with-python), !windows"
```
Comment (Neumann-A) [inline ports/libxslt/vcpkg.json:27]: Why `!windows`? Some of the code even has ifdefs for `_MSC_VER`

File: ports/libxslt/portfile.cmake [hunk 2]
```diff
@@ -125,13 +125,29 @@ if (VCPKG_TARGET_IS_WINDOWS)
         file(COPY "${CURRENT_PACKAGES_DIR}/lib/pkgconfig" DESTINATION "${CURRENT_PACKAGES_DIR}/debug/lib")
     endif()
 else()
+    if("python" IN_LIST FEATURES)
```
Comment (JonLiu1993) [inline ports/libxslt/portfile.cmake:128]: @GordonSmith ,Thanks for your pr, I think this should be a triplet variable, like here:
https://github.com/microsoft/vcpkg/pull/22409
This is vcpkg maintainer-guide we don't add options as features
https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#do-not-use-features-to-control-alternatives-in-published-interfaces

Comment (dg0yt) [reply ports/libxslt/portfile.cmake:128]: @JonLiu1993 Of course feature are for options (that are additive). They are not for alternatives (one option excluding the other). So the python feature is correct.

File: ports/libxslt/vcpkg.json (hunk 1 above)
Comment (GordonSmith) [reply ports/libxslt/vcpkg.json]: It was left over from initial implementation in my local environment - removed.

**Comment by JonLiu1993** [comment]
@GordonSmith ,Thanks for your pr, have you tested the features locally?

**Comment by GordonSmith** [comment]
@JonLiu1993 - I am using it for Win32/64, Ubuntu 20.04 and OSX 11 builds with the following manifest:
```
        {
            "name": "libxslt",
            "default-features": false,
            "features": []
        },
```
The default features are the same as  prior to this PR.
I have not tested with the Python features enabled...

**Review by JonLiu1993 — APPROVED** [review]

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

File: ports/libxslt/vcpkg.json [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
-  ]
+  ],
+  "default-features": [
+    "crypto"
+  ],
+  "features": {
+    "crypto": {
+      "description": "Build with crypto support (--with-crypto)",
+      "dependencies": [
+        "libgcrypt"
```
[the remark this answers is not in the document]
Comment (strega-nil-ms) [reply ports/libxslt/vcpkg.json:20]: It also builds without the dependency on arm64-osx...

[the remark this answers is not in the document]
Comment (GordonSmith) [reply ports/libxslt/vcpkg.json:20]: Its bit of a catch 22 I think (and a limitation in the manifest capabilities).

Prior to this PR, libxslt wouldn't build on my Windows, Linux or OSX environments.  For Linux and OSX I needed to add the libgcrypt dependency. But for Windows I needed to remove the hard coded "--with-crypto" option.

My original PR had the following descriptions:
```
      "description": "Build with crypto support (--with-crypto), !windows",
      "dependencies": [
        "libgcrypt"
      ]
    },
    "plugins": {
      "description": "Build with plugins (--with-plugins), !windows"
    },
    "python": {
      "description": "Builds with python support (--with-python), !windows"
``` 
Which was my attempt to articulate what I had observed.

For my actual use case I only needed a way to disable the "--with-crypto" option, but thought the above may be beneficial to others?

Let me know what the consensus is and I will make the modifications...

[the remark this answers is not in the document]
Comment (GordonSmith) [reply ports/libxslt/vcpkg.json:20]: @strega-nil-ms WRT to the arm64-OSX building without the dependency only indicates that the build machine in question had the dependency already installed via the OS or some other package manager?

I am not 100% sure what the correct approach is for the above scenario (if there is an official policy I would love to know what it is)?

**Comment by strega-nil-ms** [comment]
@GordonSmith could you merge the following patch into your code?

```patch
From a22c233fd0fdfc6ade9ce50409228669354596d0 Mon Sep 17 00:00:00 2001
From: nicole mazzuca <mazzucan@outlook.com>
Date: Thu, 21 Apr 2022 15:54:47 -0700
Subject: [PATCH] nicole CRs

---
 0001-nicole-CRs.patch                    | 240 +++++++++++++++++++++++
 ports/libxslt/portfile.cmake             |   5 +-
 ports/libxslt/require-dependencies.patch |  44 +++++
 ports/libxslt/vcpkg.json                 |  21 +-
 versions/l-/libxslt.json                 |   2 +-
 5 files changed, 308 insertions(+), 4 deletions(-)
 create mode 100644 0001-nicole-CRs.patch
 create mode 100644 ports/libxslt/require-dependencies.patch

diff --git a/0001-nicole-CRs.patch b/0001-nicole-CRs.patch
new file mode 100644
index 000000000..af22eb6ea
--- /dev/null
+++ b/0001-nicole-CRs.patch
@@ -0,0 +1,240 @@
+From efd4168c13c0eccfce837f5bb2cb7bad64e132b3 Mon Sep 17 00:00:00 2001
+From: nicole mazzuca <mazzucan@outlook.com>
+Date: Thu, 21 Apr 2022 15:54:47 -0700
+Subject: [PATCH] nicole CRs
+
+---
+ 0001-nicole-CRs.patch                    | 86 ++++++++++++++++++++++++
+ ports/libxslt/portfile.cmake             |  5 +-
+ ports/libxslt/require-dependencies.patch | 44 ++++++++++++
+ ports/libxslt/vcpkg.json                 | 21 +++++-
+ versions/l-/libxslt.json                 |  2 +-
+ 5 files changed, 154 insertions(+), 4 deletions(-)
+ create mode 100644 0001-nicole-CRs.patch
+ create mode 100644 ports/libxslt/require-dependencies.patch
+
+diff --git a/0001-nicole-CRs.patch b/0001-nicole-CRs.patch
+new file mode 100644
+index 000000000..3422699a7
+--- /dev/null
++++ b/0001-nicole-CRs.patch
+@@ -0,0 +1,86 @@
++From 67daed588f1f51c28beee4fe83a72ae6cdb81c8c Mon Sep 17 00:00:00 2001
++From: nicole mazzuca <mazzucan@outlook.com>
++Date: Thu, 21 Apr 2022 15:54:47 -0700
++Subject: [PATCH] nicole CRs
++
++---
++ ports/libxslt/portfile.cmake |  4 +++-
++ ports/libxslt/vcpkg.json     | 21 +++++++++++++++++++--
++ versions/l-/libxslt.json     |  2 +-
++ 3 files changed, 23 insertions(+), 4 deletions(-)
++
++diff --git a/ports/libxslt/portfile.cmake b/ports/libxslt/portfile.cmake
++index dd9c9db5c..35c6b67f1 100644
++--- a/ports/libxslt/portfile.cmake
+++++ b/ports/libxslt/portfile.cmake
++@@ -126,7 +126,8 @@ if (VCPKG_TARGET_IS_WINDOWS)
++     endif()
++ else()
++     if("python" IN_LIST FEATURES)
++-        set(OPTION_PYTHON "--with-python")
+++        vcpkg_find_acquire_program(PYTHON3)
+++        set(OPTION_PYTHON "--with-python=${PYTHON3}")
++     else()
++         set(OPTION_PYTHON "--without-python")
++     endif()
++@@ -136,6 +137,7 @@ else()
++         set(OPTION_PLUGINS "--without-plugins")
++     endif()
++     if("crypto" IN_LIST FEATURES)
+++        vcpkg_add_to_path(PREPEND "${CURRENT_INSTALLED_DIR}/tools/libgcrypt/bin")
++         set(OPTION_CRYPTO "--with-crypto")
++     else()
++         set(OPTION_CRYPTO "--without-crypto")
++diff --git a/ports/libxslt/vcpkg.json b/ports/libxslt/vcpkg.json
++index ea561e8cd..1912a492f 100644
++--- a/ports/libxslt/vcpkg.json
+++++ b/ports/libxslt/vcpkg.json
++@@ -11,13 +11,30 @@
++     "libxml2"
++   ],
++   "default-features": [
++-    "crypto"
+++    "default-features"
++   ],
++   "features": {
++     "crypto": {
++       "description": "Build with crypto support (--with-crypto)",
+++      "supports": "mingw | !windows",
++       "dependencies": [
++-        "libgcrypt"
+++        {
+++          "name": "libgcrypt",
+++          "platform": "!mingw"
+++        }
+++      ]
+++    },
+++    "default-features": {
+++      "description": "default features for the current platform",
+++      "dependencies": [
+++        {
+++          "name": "libxslt",
+++          "default-features": false,
+++          "features": [
+++            "crypto"
+++          ],
+++          "platform": "mingw | !windows"
+++        }
++       ]
++     },
++     "plugins": {
++diff --git a/versions/l-/libxslt.json b/versions/l-/libxslt.json
++index 7bc54e3c2..5875932ae 100644
++--- a/versions/l-/libxslt.json
+++++ b/versions/l-/libxslt.json
++@@ -1,7 +1,7 @@
++ {
++   "versions": [
++     {
++-      "git-tree": "d09c74ff3cdb46f6cd77cd0136c6814f4ed2cb01",
+++      "git-tree": "880b2355e9eccc8dcc89ba2618a94248c61017a0",
++       "version": "1.1.34",
++       "port-version": 5
++     },
++-- 
++2.32.0 (Apple Git-132)
++
+diff --git a/ports/libxslt/portfile.cmake b/ports/libxslt/portfile.cmake
+index dd9c9db5c..3e8917227 100644
+--- a/ports/libxslt/portfile.cmake
++++ b/ports/libxslt/portfile.cmake
+@@ -20,6 +20,7 @@ vcpkg_from_github(
+         0002-Fix-lzma.patch
+         0003-Fix-configure.patch
+         only_build_one_lib_type.patch
++        require-dependencies.patch
+ )
+ 
+ if (VCPKG_TARGET_IS_WINDOWS)
+@@ -126,7 +127,8 @@ if (VCPKG_TARGET_IS_WINDOWS)
+     endif()
+ else()
+     if("python" IN_LIST FEATURES)
+-        set(OPTION_PYTHON "--with-python")
++        vcpkg_find_acquire_program(PYTHON3)
++        set(OPTION_PYTHON "--with-python=${PYTHON3}")
+     else()
+         set(OPTION_PYTHON "--without-python")
+     endif()
+@@ -136,6 +138,7 @@ else()
+         set(OPTION_PLUGINS "--without-plugins")
+     endif()
+     if("crypto" IN_LIST FEATURES)
++        vcpkg_add_to_path(PREPEND "${CURRENT_INSTALLED_DIR}/tools/libgcrypt/bin")
+         set(OPTION_CRYPTO "--with-crypto")
+     else()
+         set(OPTION_CRYPTO "--without-crypto")
+diff --git a/ports/libxslt/require-dependencies.patch b/ports/libxslt/require-dependencies.patch
+new file mode 100644
+index 000000000..8bf853665
+--- /dev/null
++++ b/ports/libxslt/require-dependencies.patch
+@@ -0,0 +1,44 @@
++--- a/configure.ac
+++++ b/configure.ac
++@@ -240,12 +240,13 @@
++     else
++ 	if test -x "$with_python"
++ 	then
++ 	    echo Found python in $with_python
++ 	    PYTHON="$with_python"
++ 	else
+++            AC_MSG_ERROR([Python not found; --with-python should be set to the python path])
++             if test -x "$PYTHON"
++             then
++                 echo Found python in environment PYTHON=$PYTHON
++                 with_python=`$PYTHON -c "import sys; print sys.exec_prefix"`
++ 	    else
++ 	        AC_PATH_PROG(PYTHON, python python2.4 python2.3 python2.2 python2.1 python2.0 python1.6 python1.5)
++@@ -333,25 +334,25 @@
++     if test "$LIBGCRYPT_CONFIG" != "no" ; then
++       LIBGCRYPT_VERSION=`$LIBGCRYPT_CONFIG --version`
++       if test VERSION_TO_NUMBER(echo $LIBGCRYPT_VERSION) -lt VERSION_TO_NUMBER(echo "1.1.42")
++       then
++         LIBGCRYPT_CFLAGS=""
++         LIBGCRYPT_LIBS=""
++-        echo 'gcrypt library version < 1.1.42 - Crypto extensions will not be available.'
+++        AC_MSG_ERROR([gcrypt library version < 1.1.42 - this is an error])
++       else
++         LIBGCRYPT_CFLAGS=`$LIBGCRYPT_CONFIG $libgcrypt_config_args --cflags`
++         LIBGCRYPT_LIBS=`$LIBGCRYPT_CONFIG $libgcrypt_config_args --libs`
++         AC_DEFINE(HAVE_GCRYPT, 1, [Define if gcrypt library is available.])
++         echo 'Crypto extensions will be available.'
++         WITH_CRYPTO=1
++         CRYPTO_TESTDIR=crypto
++       fi
++     else
++       LIBGCRYPT_CFLAGS=""
++       LIBGCRYPT_LIBS=""
++-      echo 'Crypto extensions will not be available. Install libgcrypt and reconfigure to make available.'
+++      AC_MSG_ERROR([libgcrypt not found - this is a bug in the vcpkg port])
++     fi
++ esac
++ fi
++ AC_SUBST(WITH_CRYPTO)
++ AC_SUBST(CRYPTO_TESTDIR)
++ AC_SUBST(LIBGCRYPT_CFLAGS)
+diff --git a/ports/libxslt/vcpkg.json b/ports/libxslt/vcpkg.json
+index ea561e8cd..1912a492f 100644
+--- a/ports/libxslt/vcpkg.json
++++ b/ports/libxslt/vcpkg.json
+@@ -11,13 +11,30 @@
+     "libxml2"
+   ],
+   "default-features": [
+-    "crypto"
++    "default-features"
+   ],
+   "features": {
+     "crypto": {
+       "description": "Build with crypto support (--with-crypto)",
++      "supports": "mingw | !windows",
+       "dependencies": [
+-        "libgcrypt"
++        {
++          "name": "libgcrypt",
++          "platform": "!mingw"
++        }
++      ]
++    },
++    "default-features": {
++      "description": "default features for the current platform",
++      "dependencies": [
++        {
++          "name": "libxslt",
++          "default-features": false,
++          "features": [
++            "crypto"
++          ],
++          "platform": "mingw | !windows"
++        }
+       ]
+     },
+     "plugins": {
+diff --git a/versions/l-/libxslt.json b/versions/l-/libxslt.json
+index 7bc54e3c2..5875932ae 100644
+--- a/versions/l-/libxslt.json
++++ b/versions/l-/libxslt.json
+@@ -1,7 +1,7 @@
+ {
+   "versions": [
+     {
+-      "git-tree": "d09c74ff3cdb46f6cd77cd0136c6814f4ed2cb01",
++      "git-tree": "880b2355e9eccc8dcc89ba2618a94248c61017a0",
+       "version": "1.1.34",
+       "port-version": 5
+     },
+-- 
+2.32.0 (Apple Git-132)
+
diff --git a/ports/libxslt/portfile.cmake b/ports/libxslt/portfile.cmake
index dd9c9db5c..3e8917227 100644
--- a/ports/libxslt/portfile.cmake
+++ b/ports/libxslt/portfile.cmake
@@ -20,6 +20,7 @@ vcpkg_from_github(
         0002-Fix-lzma.patch
         0003-Fix-configure.patch
         only_build_one_lib_type.patch
+        require-dependencies.patch
 )
 
 if (VCPKG_TARGET_IS_WINDOWS)
@@ -126,7 +127,8 @@ if (VCPKG_TARGET_IS_WINDOWS)
     endif()
 else()
     if("python" IN_LIST FEATURES)
-        set(OPTION_PYTHON "--with-python")
+        vcpkg_find_acquire_program(PYTHON3)
+        set(OPTION_PYTHON "--with-python=${PYTHON3}")
     else()
         set(OPTION_PYTHON "--without-python")
     endif()
@@ -136,6 +138,7 @@ else()
         set(OPTION_PLUGINS "--without-plugins")
     endif()
     if("crypto" IN_LIST FEATURES)
+        vcpkg_add_to_path(PREPEND "${CURRENT_INSTALLED_DIR}/tools/libgcrypt/bin")
         set(OPTION_CRYPTO "--with-crypto")
     else()
         set(OPTION_CRYPTO "--without-crypto")
diff --git a/ports/libxslt/require-dependencies.patch b/ports/libxslt/require-dependencies.patch
new file mode 100644
index 000000000..8bf853665
--- /dev/null
+++ b/ports/libxslt/require-dependencies.patch
@@ -0,0 +1,44 @@
+--- a/configure.ac
++++ b/configure.ac
+@@ -240,12 +240,13 @@
+     else
+ 	if test -x "$with_python"
+ 	then
+ 	    echo Found python in $with_python
+ 	    PYTHON="$with_python"
+ 	else
++            AC_MSG_ERROR([Python not found; --with-python should be set to the python path])
+             if test -x "$PYTHON"
+             then
+                 echo Found python in environment PYTHON=$PYTHON
+                 with_python=`$PYTHON -c "import sys; print sys.exec_prefix"`
+ 	    else
+ 	        AC_PATH_PROG(PYTHON, python python2.4 python2.3 python2.2 python2.1 python2.0 python1.6 python1.5)
+@@ -333,25 +334,25 @@
+     if test "$LIBGCRYPT_CONFIG" != "no" ; then
+       LIBGCRYPT_VERSION=`$LIBGCRYPT_CONFIG --version`
+       if test VERSION_TO_NUMBER(echo $LIBGCRYPT_VERSION) -lt VERSION_TO_NUMBER(echo "1.1.42")
+       then
+         LIBGCRYPT_CFLAGS=""
+         LIBGCRYPT_LIBS=""
+-        echo 'gcrypt library version < 1.1.42 - Crypto extensions will not be available.'
++        AC_MSG_ERROR([gcrypt library version < 1.1.42 - this is an error])
+       else
+         LIBGCRYPT_CFLAGS=`$LIBGCRYPT_CONFIG $libgcrypt_config_args --cflags`
+         LIBGCRYPT_LIBS=`$LIBGCRYPT_CONFIG $libgcrypt_config_args --libs`
+         AC_DEFINE(HAVE_GCRYPT, 1, [Define if gcrypt library is available.])
+         echo 'Crypto extensions will be available.'
+         WITH_CRYPTO=1
+         CRYPTO_TESTDIR=crypto
+       fi
+     else
+       LIBGCRYPT_CFLAGS=""
+       LIBGCRYPT_LIBS=""
+-      echo 'Crypto extensions will not be available. Install libgcrypt and reconfigure to make available.'
++      AC_MSG_ERROR([libgcrypt not found - this is a bug in the vcpkg port])
+     fi
+ esac
+ fi
+ AC_SUBST(WITH_CRYPTO)
+ AC_SUBST(CRYPTO_TESTDIR)
+ AC_SUBST(LIBGCRYPT_CFLAGS)
diff --git a/ports/libxslt/vcpkg.json b/ports/libxslt/vcpkg.json
index ea561e8cd..1912a492f 100644
--- a/ports/libxslt/vcpkg.json
+++ b/ports/libxslt/vcpkg.json
@@ -11,13 +11,30 @@
     "libxml2"
   ],
   "default-features": [
-    "crypto"
+    "default-features"
   ],
   "features": {
     "crypto": {
       "description": "Build with crypto support (--with-crypto)",
+      "supports": "mingw | !windows",
       "dependencies": [
-        "libgcrypt"
+        {
+          "name": "libgcrypt",
+          "platform": "!mingw"
+        }
+      ]
+    },
+    "default-features": {
+      "description": "default features for the current platform",
+      "dependencies": [
+        {
+          "name": "libxslt",
+          "default-features": false,
+          "features": [
+            "crypto"
+          ],
+          "platform": "mingw | !windows"
+        }
       ]
     },
     "plugins": {
diff --git a/versions/l-/libxslt.json b/versions/l-/libxslt.json
index 7bc54e3c2..f1d753a3e 100644
--- a/versions/l-/libxslt.json
+++ b/versions/l-/libxslt.json
@@ -1,7 +1,7 @@
 {
   "versions": [
     {
-      "git-tree": "d09c74ff3cdb46f6cd77cd0136c6814f4ed2cb01",
+      "git-tree": "063e2e6a6b9ef700f0c2587d2def57eb0e2285e6",
       "version": "1.1.34",
       "port-version": 5
     },
-- 
2.32.0 (Apple Git-132)
```

**Review by strega-nil-ms — APPROVED** [review]

**Comment by strega-nil-ms** [comment]
Thanks!

**Pull request merged**